### DISCUSSION TITLE
How to select the models to use?

### DISCUSSION THREAD

**Discussion by haimat** [text2knowledge/docTR-Labeler#5] [General]
I would like to use docTR-Labeler to annoate our custom dataset. Is there a way to specify which detection and recognition models the application should use for the AUTO ANNOTATE button?

**Comment by felixdittrich92** [comment]
Not yet with other models as `fast_` there seems to be an issue I have to fix before.

The easiest would be to write a quick script on your own:

For example if you download [reference datasets](https://drive.google.com/file/d/1brbktJZO4r1jR9U8JAGJUp7mM7Ctn4gI/view?usp=sharing)

and inside the `detection_task/train folder` rename the `labels.json` to `pre_annotations.json` and open the `detection_task/train/images` folder with the tool you will see that the annotations are imported correctly.

So everything you would need to do is to pre-label your data with docTR and save it in the correct format (`images` folder and the corresponding `pre_annotations.json`) then you can open it with the tool and correct.

**NOTE: Also already complete correct pre annotated images needs to be saved inside the tool otherwise the merged file after all images are done at the end is incomplete** !

**Reply by felixdittrich92** [reply]
ok I found the bug let me fix this quickly :sweat_smile:

**Reply by felixT2K** [reply]
@haimat I did a new release (v0.1.4)

Now you can specify before starting the tool which models it should use:
https://github.com/text2knowledge/docTR-Labeler#configuration

**Reply by felixT2K** [reply]
one tip:

I suggest to use the "tighten" functionality we provide - this will later while training improve the results you get highly because the polygon annotations are much closer to the words

**Reply by haimat** [reply]
Thank you very much for this update, Felix. Maybe I am doining something wrong, but it seems this command does not change anything on my end with 0.1.4:

    RECOGNITION_ARCH="crnn_mobilenet_v3_large" DETECTION_ARCH="db_mobilenet_v3_large" doctr-labeler

Any ideas?

**Reply by felixdittrich92** [reply]
Mh ..tested and it works on my end ..

```
(doctr-labeler) felix@felix-Z790-AORUS-MASTER:~/Desktop/docTR-Labeler$ RECOGNITION_ARCH="crnn_mobilenet_v3_large" DETECTION_ARCH="db_mobilenet_v3_large" doctr-labeler
db_mobilenet_v3_large
crnn_mobilenet_v3_large
```

**Reply by felixdittrich92** [reply]
Could you try to reduce the threshold also please:

```
RECOGNITION_ARCH=crnn_mobilenet_v3_large DETECTION_ARCH=db_mobilenet_v3_large OBJECTNESS_THRESHOLD=0 doctr-labeler
```

**Reply by haimat** [reply]
Still not difference. I know that the specified model is not used, because with in the labeler tool I see the wrong text characters recgonized, whereas when I run the same recognition model as specified on the command line directly on the same image (thus outside the labeler tool), then the recognized characters are correct.

In other words: The same model applied directly to an image works, but in the labeler tool when specified on the command line leads to a different result.

**Reply by felixT2K** [reply]
Ah I think I know the issue .. In the tool we use [OnnxTR](https://github.com/felixdittrich92/OnnxTR) (which is the same as docTR at the end but doesn't requires TensorFlow or PyTorch - works with onnxruntime under the hood)
While exporting the torch models to onnx it's possible that a really small numerical difference exists - in some small edge cases this can lead to different results

**Reply by felixT2K** [reply]
The recognition issue comes after your draw a box on your own or after "auto annotation" ?

**Reply by haimat** [reply]
It's after auto-annotation. 
Would it be difficult to run the label tool with docTR (instead of onnxTR)?

**Reply by haimat** [reply]
I have the impression even though I specify the models to use on command line, they are not being used in the label tool. Because with or without the models specified, the auto-generated bounding boxes look exactly the same. As if in both cases the same models are being used internally. 

Maybe to verify we can print the models in use to console when the program starts?

**Reply by felixdittrich92** [reply]
Yeah I added logging to this step you can install the repo's main branch directly with `pip3 install doctr-labeler@git+https://github.com/text2knowledge/docTR-Labeler`

**Reply by haimat** [reply]
Thanks, that did the trick for me to debug this issue.
For some reason the command line has loaded doctr-labeler from the user's pip installation folder.
I had to restart the venv in order to run the version I installed there.
Now the models from command line are being loaded correctly :+1:

**Reply by felixdittrich92** [reply]
Ah ok :+1:  Does this also fix the issue now ? :)

**Reply by haimat** [reply]
Yes, thanks, looks better now.
Next I will annotate our dataset and fine-tune one of your models to test :+1:

**Reply by felixdittrich92** [reply]
Nice glad to hear :) :+1:

**Reply by haimat** [reply]
Thanks for the quick help, I appreciate that very much!
